File: src/algorithms/Remove_Nth_Node_From_End_of_List.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:

    def remove_node(self, head, n):
        if head.next is None:
            self.cnt += 1
            return
        else:
            self.remove_node(head.next, n)
            if self.cnt == n:
                nxt = head.next
                head.next = nxt.next
                del nxt
            self.cnt += 1

    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        self.cnt = 0
        self.remove_node(head, n)
        if self.cnt == 1:
            return None
        elif n == self.cnt:
            return head.next
        else:
            return head


if __name__ == '__main__':
    solution = Solution()
    a, b = ListNode(1), ListNode(2)
    c, d = ListNode(3), ListNode(4)
    e = ListNode(5)
    a.next, b.next = b, c
    c.next, d.next = d, e
    head = solution.removeNthFromEnd(a, 5)
    while head:
        print head.val
        head = head.next
    a = ListNode(1)
    head = solution.removeNthFromEnd(a, 1)
    while head:
        print head.val
        head = head.next

File: src/algorithms/cpp/Remove_Nth_Node_From_End_of_List.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
private:
    ListNode *ret;
    int n;

private:
    int dfs(ListNode *pre, ListNode *now) {
        if (now == NULL)   return 0;
        int nxt_ret = dfs(now, now->next);
        if (nxt_ret == n - 1) {
            if (pre == NULL)    this->ret = now->next;
            else {
                pre->next = now->next;
                free(now);
            }
        }
        return nxt_ret + 1;
    }

public:
    ListNode* removeNthFromEnd(ListNode *head, int n) {
        this->ret = head;
        this->n = n;
        dfs(NULL, head);
        return this->ret;
    }
};

int main() {
    return 0;
}
